### PR TITLE
Feature/pfconfig sync pp

### PR DESCRIPTION
**Pull request by julsemaan**
# Description

Adds better error propagation on sync failure in bin/cluster/sync
Admin will show a failure when saving and the cluster fails to sync completely
Adds an option to fix permissions of specific files in `pfcmd fixpermissions`
Fix the permissions of a file before synching so pf can write to it
# Impacts

Active/active sync
# Delete branch after merge

YES
# NEWS file entries
## Enhancements
- Added better error message propagation during a cluster synchronization
- pfcmd fixpermissions can not be used on specific files
## Bug Fixes
- Fix bad permissions before synching a file in a cluster

### CODE REVIEW

File: lib/pf/cmd/pf/fixpermissions.pm [hunk 1]
```diff
@@ -18,19 +18,20 @@ pf::cmd::pf::fixpermissions
 use strict;
```
Comment (rouzier) [inline lib/pf/cmd/pf/fixpermissions.pm:21]: Update pod doc to reflect new options

File: lib/pf/cmd/pf/fixpermissions.pm [hunk 2]
```diff
@@ -39,6 +40,16 @@ sub _run {
     chmod(0664, @stored_config_files);
     chmod(02775, $conf_dir, $var_dir, $log_dir);
     _fingerbank();
+    print "Fixed permissions.\n";
+    return $EXIT_SUCCESS;
+}
+
+sub action_file {
```
Comment (jrouzierinverse) [inline lib/pf/cmd/pf/fixpermissions.pm:79]: Create a parse_file method to check if the file exists

File: lib/pf/api.pm [hunk 3]
```diff
@@ -480,10 +480,11 @@ sub notify_configfile_changed : Public {
 
     my $apiclient = pf::api::jsonrpcclient->new(proto => 'https', host => $master_server->{management_ip});
 
+    pf::util::pf_run("sudo /usr/local/pf/bin/pfcmd fixpermissions file $postdata{conf_file}");
+    open(my $fh, '>', $postdata{conf_file}) or die "Cannot open file $postdata{conf_file} for writing.'";
```
Comment (jrouzierinverse) [inline lib/pf/api.pm]: Should re-throw error below

**Comment by jrouzierinverse** [comment]
As a safety measure we should only allow the api to modify/retrieve allowed files in conf directories.

File: lib/pf/cmd/pf/fixpermissions.pm (hunk 2 above)
Comment (jrouzierinverse) [inline lib/pf/cmd/pf/fixpermissions.pm:79]: It would be nice to accept multiple files.

File: lib/pf/cmd/pf/fixpermissions.pm [hunk 4]
```diff
@@ -39,6 +40,16 @@ sub _run {
...
+    my ($self) = @_;
+    my ($file) = $self->action_args;
+    $file = untaint_chain($file);
```
Comment (jrouzierinverse) [inline lib/pf/cmd/pf/fixpermissions.pm]: Check to see if the file path is in /usr/local/pf or /usr/local/fingerbank

File: lib/pf/api.pm (hunk 3 above)
Comment (julsemaan) [inline lib/pf/api.pm]: Why ? It's already going up the chain and generating a 500

File: lib/pf/cmd/pf/fixpermissions.pm [hunk 5]
```diff
@@ -39,6 +43,52 @@ sub _run {
...
+        }
+        unless($file =~ /\/usr\/local\/pf\// || $file =~ /\/usr\/local\/fingerbank\//){
+            print STDERR "File $file is not in an allowed directory \n";
+            return 0;
+        }
+    }
+    return 1;
+}
+
+sub action_file {
```
Comment (jrouzierinverse) [inline lib/pf/cmd/pf/fixpermissions.pm:79]: doc

File: lib/pf/cmd/pf/fixpermissions.pm [hunk 6]
```diff
@@ -18,19 +21,20 @@ pf::cmd::pf::fixpermissions
...
 
 use fingerbank::FilePath;
 
 use File::Spec::Functions qw(catfile);
 
-sub parseArgs { 1 }
+sub default_action { 'all' }
 
-sub _run {
+sub action_all {
```
Comment (jrouzierinverse) [inline lib/pf/cmd/pf/fixpermissions.pm:43]: doc

**Comment by julsemaan** [comment]
Re-review

**Pull request merged**